### PR TITLE
Goal metadata

### PR DESCRIPTION
**Pull request by robdockins**
This PR improves the metadata attached to goals, and implements a system whereby user-defined "tags" can be associated with specification conditions in the setup phase. Later, in proof tactics, goals can be filtered by what tags they have associated. This should give users a less-fragile way to filter goals for special handling.

The system may need additional refinement going forward, but this seems like  good step in the right direction.  One thing I notice is that a single specification statement can produce multiple goals (e.g., a points-to requires both that the a memory read succeeds, and that the resulting value is the one expected), each of which have the same tags.

### CODE REVIEW

File: src/SAWScript/X86.hs [hunk 1]
```diff
@@ -405,6 +409,9 @@ translate opts elf fun =
   do let name = funName fun
      sayLn ("Translating function: " ++ BSC.unpack name)
 
+     -- TODO? do we need to pass in the mdMap into more places in this mode?
```
Comment (chameco) [inline src/SAWScript/X86.hs:412]: I'm also unsure of this - AFAIK there's only one old proof currently using this mode, at https://github.com/GaloisInc/aes-gcm-siv-proof. This builds against an old SAW commit anyway. There's probably a conversation to be had about deprecating/removing this interface.

**Review by chameco — APPROVED** [review]
Looks good overall, this is exciting. We should open an issue detailing any missing JVM functionality.

File: src/SAWScript/X86.hs (hunk 1 above)
Comment (robdockins) [reply src/SAWScript/X86.hs:412]: If this isn't an interface that we intend to use for future proofs, then I probably won't worry to much about piping in new usability features.

Is proof something that, in principle, can be done with other methods now?

Comment (robdockins) [reply src/SAWScript/X86.hs:412]: OK, in that case, I'll leave this alone, and it probably is worth having a discussion about removing this interface.

**Pull request merged**